---
title: "Decode Intercellular Communication by Computational Method-NicheNet"
author: "Jia Zhao"
date: "2024-04-02"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Installation of nichenetr and load packages

```{r}
install.packages("devtools")
devtools::install_github("saeyslab/nichenetr")

# You might need to install dependencies to successfully install nichenetr
# install.packages("BiocManager")
# BiocManager::install("limma", force = TRUE)
# BiocManager::install("ComplexHeatmap", force = TRUE)

# Load nichenetr and tidyverse
library(nichenetr)

#install.packages("tidyverse")
# tidyverse includes packages like ggplot2 for data visualization, dplyr for data manipulation, readr for reading data, and several others
library(tidyverse)

.
```

**NicheNet can predict:**

1)  **Upstream ligand analysis**: which ligands from one cell population (“sender/niche”) are most likely to affect target gene expression in an interacting cell population (“receiver/target”). This requires prior knowledge of defining a gene set of interest in the receiver cells (e.g. control vs mutant -\> find differentially expressed genes).

2)  **Downstream ligand activity analysis**: which specific target genes are affected by which of these predicted ligands.

3)  Other functionalities include:

    -   assessing how well ligands expressed by a sender cell can predict changes in gene expression in the receiver cell

    -   prioritizing ligands based on their effect on gene expression

    -   inferring putative ligand-target links active in the system under study

    -   inferring potential signaling paths between ligands and target genes of interest: to generate causal hypotheses and check which data sources support the predictions

    -   validation of the prior ligand-target model

    -   construction of user-defined prior ligand-target model

# Problem definition:

Tissue resident macrophages (TRMs) are present ubiquitously in every tissue and organ, yet how they support tissue and organ function is largely unknown. Since intercellular commnunication is critical for tissue homeostasis, we aim to study intercellular commnunication by infer interaction between TRMs and other cell types.

This document is focused on the infering intercellular communication between TRMs and the primary cells adipocytes in white adipose tissue. Since the paper calls fat TRMs as VAMs, we will use VAMs from now on.

**Workflow:**

1.  Define a sender cell population and a receiver cell population and determine which genes are expressed in both populations:

    *a. sender: VAMs, receiver: adipocytes*

2.  (**Critical step**) Define a gene set. **Define differential expressed genes (DEGs) in receiver cells potentially affected by ligands** (*Most challenging*)

    a\. DEGs: Use adipocytes RNA-seq data vs preadipocytes (called FRCs (fibroblast reticular cells) in Moura Silva dataset)

3.  Define a ligand set: from sender cells and can potentially bind to receiver cells

4.  NicheNet ligand activity analysis: rank ligands based on the gene expression of target genes in the receiver cells

5.  Infer top-predicted target genes based on top-ranked ligands

# Step 0: Load NicheNet’s ligand-target prior model and processed expression data of interacting cells

```{r}
options(timeout = 600)
organism = "mouse"

if(organism == "human"){
  lr_network = readRDS(url("https://zenodo.org/record/7074291/files/lr_network_human_21122021.rds"))
  ligand_target_matrix = readRDS(url("https://zenodo.org/record/7074291/files/ligand_target_matrix_nsga2r_final.rds"))
} else if(organism == "mouse"){
  lr_network = readRDS(url("https://zenodo.org/record/7074291/files/lr_network_mouse_21122021.rds"))
  ligand_target_matrix = readRDS(url("https://zenodo.org/record/7074291/files/ligand_target_matrix_nsga2r_final_mouse.rds"))

}

lr_network = lr_network %>% distinct(from, to)
ligand_target_matrix[1:5,1:5] # target genes in rows, ligands in columns

```

**Data**: Expression data of interacting cells: publicly available RNA-seq data in white adipose tissue, including VAMs and adipocytes from wild-type adult C57BL/6J male mice (12-16 wk):

**Motivation**: Macrophage can sense the environment and coordinate with other cells. It is intriguing that VAMs are present ubiquitously inside the white adipose tissue. What are their functions? Surprisingly, VAMs are present in neonate fat before mature adipocyte, indicating the role of VAMs in adipocyte differentiation and maturation. How do VAMs communicate with adipocytes?

**Hypothesis**: VAMs-ligands regulate the expression of adipocytes-maturation genes

**Import our RNA-seq dataset containing TRM subset named VAM1 (Vasculature Associated Macrophages 1) and adipocytes for WT and c-Maf deletion**

```{r}
# Need to adjust to your download file! Named: Silva_WAT_cMAF_vs_WT.csv
# Since I download
cMaf_data <- read_csv('~/Downloads/Silva_WAT_cMAF_vs_WT.csv')
head(cMaf_data)
```

```{r}
# Because the NicheNet 2.0. networks are in the most recent version of the official gene symbols, we will make sure that the gene symbols used in the expression data are also updated (= converted from their “aliases” to official gene symbols).

gene_name <- cMaf_data$gene
gene_name_symbol <- convert_alias_to_symbols(gene_name, "mouse", verbose = FALSE) %>% unique()
```

# Step 1: Define expressed genes in sender and receiver cell populations

Question: which ligands expressed by VAMs can induce adipocyte maturation from preadipocytes?

```{r}
# Make a function to extract expressuion data from our RNA-seq data set based on cell types

extract_expression <- function(data, pattern, cutoff = 100) {
  # Identify columns matching the pattern: e.g. "VAM1-WT", "Adipocyte-WT"
  expression_oi = grepl(pattern, names(data)) | grepl("gene", names(data))
  
  # Include the first column explicitly
  expression_oi[1] = TRUE
  
  # Extract relevant columns
  expression_df = data[, expression_oi]
  
  # Assume the gene expression columns start from the 2nd column to 4th column
  expression_df$average = rowMeans(expression_df[, 2:4], na.rm = TRUE)
  
  # Filter based on the cutoff
  filtered_expression_df = expression_df[expression_df$average > cutoff, ]
  
  # Print number of rows after filtering
  nrow <- nrow(filtered_expression_df) %>% unique()
  print(nrow)
  
  return(filtered_expression_df)
}
```

## Sender cells: VAMs

```{r}
VAMs_expression_df <- extract_expression(data = cMaf_data, pattern = "VAM1-WT", cutoff = 100)

expressed_genes_sender <-VAMs_expression_df
```

## Receiver cells: Adipocytes/FRCs

```{r}
# Extract receiver cells expression: Adipocytes and adipocytes progeneitor (FRCs) 

# Adipocytes
Adipocyte_expression_df <- extract_expression(data = cMaf_data, pattern = "Adipocyte-WT", cutoff = 100)
# number of rows of the filtered data
# [1] 9905


# FRCs
FRC_expression_df <- extract_expression(data = cMaf_data, pattern = "FRC-WT", cutoff = 100)
# number of rows of the filtered data
# [1] 9903

expressed_genes_receiver_adipocyte <- Adipocyte_expression_df
expressed_genes_receiver_FRC <- FRC_expression_df
```

# Step 2: Define the gene set of interest and a background of genes

## DEG analysis to define potential genes involved in adipocyte maturation

Here I can start by find differentially expressed genes (DEGs) between adipocytes and preadipocytes to find genes involved in adipocyte maturation

**Note**: As gene set of interest, we consider the genes of which the expression is possibly affected due to communication with other cells. The definition of this gene set depends on your research question and is a crucial step in the use of NicheNet.

### Prepare data

```{r}
# Load packages
library(DESeq2)

# Extract relevant columns
# Ensure the gene column is not included in the counts matrix
relevant_columns <- grep("gene|Adipocyte-WT|FRC-WT", names(cMaf_data))
counts_matrix_gene <- as.matrix(cMaf_data[, relevant_columns]) 

# The counts_matrix_gene contains float data which is not compatible with DESeq2. Need to convert to integers
# Prep the matrix to remove gene name and convert to interger
counts_matrix <- counts_matrix_gene[,-1]
# Convert the counts_matrix to integers
counts_matrix <- apply(counts_matrix, 2, as.integer)

# Set row names of counts_matrix to gene names from the first column of cMaf_data
rownames(counts_matrix) <- counts_matrix_gene[, 1]

# Prepare metadata
conditions <- c(rep("Adipocyte-WT", 3), rep("FRC-WT", 3)) # 3 biological replicates
colData <- data.frame(row.names = colnames(counts_matrix), condition = conditions)


```

### DESeq2 Analysis

```{r}
# Perform DEG analysis
dds <- DESeqDataSetFromMatrix(countData = counts_matrix,
                              colData = colData,
                              design = ~ condition)
dds <- DESeq(dds)
# "Adipocyte-WT" is treated as the experimental condition, and "FRC-WT" as the reference
results <- results(dds, contrast = c("condition", "Adipocyte-WT", "FRC-WT"))

# Adjusting for multiple testing, for example, using Benjamini-Hochberg procedure
results <- results[order(results$padj),] # Ordering by adjusted p-value

# Viewing top differentially expressed genes
head(results)

# log2 fold change (MLE): condition Adipocyte-WT vs FRC-WT 
# Wald test p-value: condition Adipocyte-WT vs FRC-WT 
# DataFrame with 6 rows and 6 columns
#        baseMean log2FoldChange     lfcSE      stat       pvalue         padj
#       <numeric>      <numeric> <numeric> <numeric>    <numeric>    <numeric>
# Cidec   24004.43       10.76535  0.252205   42.6850  0.00000e+00  0.00000e+00
# Thrsp    8810.62       10.03878  0.269953   37.1871 1.10134e-302 1.21456e-298
# Scd1   294128.25        7.61268  0.208071   36.5869 4.61853e-293 3.39554e-289
# Adipoq  26110.90       11.27956  0.314089   35.9120 1.98384e-282 1.09389e-278
# Pck1    11023.12       10.94758  0.316337   34.6074 1.95692e-262 8.63236e-259
# Pnpla2   6966.05        6.17142  0.183506   33.6306 5.98560e-248 2.20031e-244

# base mean expression across samples
# log2 fold changes (indicating how much more or less a gene is expressed in one condition relative to another). Higer means more expression in Adipocyte and vice versa


```

### **Filter to identify significantly expressed genes based on three criteria**

1.  padj \< 0.05
2.  absolute log fold change between groups in normalized counts (expression) \> 2
3.  mean expression \> 100

```{r}
# Convert results to a data frame, typically needed for ggplot2-based plotting
res_df <- as.data.frame(results)

# Filter for significant genes with padj < 0.05 and absolute log fold change in normalized counts (expression) > 2 and mean expression > 100
significant_genes <- res_df[res_df$padj < 0.05 & abs(res_df$log2FoldChange) > 2 & res_df$baseMean > 100, ] %>% distinct()

nrow(significant_genes)
# [1] 1664

# Filter significantly upregulated genes & log fold change in normalized counts (expression) > 2 and rank in decreasing log2FoldCahnge
gene_Upregulated <- significant_genes[significant_genes$log2FoldChange > 2, ] %>% arrange(desc(log2FoldChange))
nrow((gene_Upregulated))
head(gene_Upregulated)
# [1] 837

# Filter significantly downregulated genes & log fold change in normalized counts (expression) < -2 and rank based on decreasing absolute value of log2FoldChange
gene_Downregulated <- significant_genes[significant_genes$log2FoldChange < -2, ] %>% arrange(desc(abs(log2FoldChange)))
nrow(gene_Downregulated)
head(gene_Downregulated)
# [1] 828
```

### Plot

**MA plot**

Visualize the log fold change of gene expression change (M) against the mean of normalized counts of the gene (A) across all samples. In this case, adipocytes and FRCs.

For y axis, positive values indicate upregulation in "WT Adipocyte" relative to "FRC", while negative values indicate downregulation.

**Points**: Each point represents a gene. Points that lie along the horizontal line at 0 indicate no change in expression between the two conditions. Points above the line are more highly expressed in "WT Adipocyte" compared to "FRC," and points below are less expressed.

```{r}
# Open a PDF device
# pdf("MA_Plot_WT_Adipocyte_vs_FRC.pdf", width = 7, height = 5)

plotMA(results, main="MA Plot: WT Adipocyte vs FRC", ylim=c(-15,15))

# Close the device to save the plot
# dev.off()
```

The basic plotMA function does not offer arguments to adjust the x axis to log scale. Need to try ggmaplot, which allows more modifications

```{r}
library(EnhancedVolcano)

# The `ggmaplot` function may require specific naming of columns; 
res_df$gene <- rownames(res_df)  # Ensure there's a 'gene' column for gene names
res_df$log2MeanExpression <- log2(res_df$baseMean + 1)  # Add log2 mean expression

# Pre-select top 15 upregulated and downregulated genes
top_upregulated_genes <- gene_Upregulated %>% arrange(desc(log2FoldChange), desc(baseMean)) %>%
  slice_head(n = 15)

top_downregulated_genes <- res_df %>%
  filter(padj < 0.05, log2FoldChange < -2) %>% arrange(desc(abs(log2FoldChange)), desc(baseMean)) %>% slice_head(n = 15)

# Combine top 15 upregulated and downregulated genes
top_genes <- c(rownames(top_upregulated_genes), top_downregulated_genes$gene)
```

```{r}
# Generate MA plot

# pdf("../figure/MA_Plot_WT_Adipocyte_vs_FRC-2.pdf", width = 7, height = 5)

# Now we'll use the ggmaplot function
ggmaplot(
  res_df,
  fdr = 0.05, # false discovery rate = 0.05
  fc = 2, # Identify genes with more than four-fold change 
  genenames = rownames(res_df),
  size = 1.5,
  alpha = 0.7,
  font.label = c(12, "plain", "black"),
  palette = c("red", "blue", "darkgray"),
  top = 20,
  select.top.method = c("padj", "fc"),
  main = "MA Plot: WT Adipocyte vs FRC",
  xlab = "Log2 mean expression",
  ylab = "Log2 fold change",
  ggtheme = theme_classic()
)

# dev.off()
```

**Volcano plot**

```{r}
# pdf("../figure/Volcano_Plot_WT_Adipocyte_vs_FRC.pdf", width = 7, height = 5)
EnhancedVolcano(res_df,
                lab = res_df$gene, # Make sure this matches your column name for gene names
                x = 'log2FoldChange',
                y = 'padj',
                xlim = c(-max(abs(res_df$log2FoldChange)), max(abs(res_df$log2FoldChange))),
                title = 'Volcano Plot: WT Adipocyte vs FRC',
                xlab = 'Log2 Fold Change',
                ylab = '-Log10 Adjusted P-value',
                selectLab = top_genes, # Genes to label: top 
                pCutoff = 0.05,
                FCcutoff = 2,
                pointSize = 2.0,
                labSize = 4.0,
                col = c('grey', 'grey', 'grey', 'blue'),
                colAlpha = 0.5,
                legendPosition = 'right',
                legendIconSize = 4.0,
                drawConnectors = TRUE,
                widthConnectors = 0.5
               )

# dev.off()
```

**Heatmap**

```{r}
library(ComplexHeatmap)
library(circlize)

# Assume 'dds' is your DESeq2 dataset and 'results' is the DESeq2 results
normalized_counts <- counts(dds, normalized=TRUE)
top_genes <- rownames(head(results[order(results$padj), ]))

# Create a named vector of colors for the conditions
condition_colors <- c("Adipocyte-WT" = "red", "FRC" = "blue")

# Make sure the condition in colData is a factor and has the correct levels
colData(dds)$condition <- factor(colData(dds)$condition, levels = names(condition_colors))

# Create the top annotation based on the condition
ha <- HeatmapAnnotation(condition = colData(dds)$condition, 
                        col = list(condition = condition_colors))

# Create the row annotation for fold changes
ra <- rowAnnotation(FoldChange = results$log2FoldChange[top_genes])

# pdf("../figure/Heatmap_Plot_WT_Adipocyte_vs_FRC.pdf", width = 7, height = 5)
# Generate the heatmap
Heatmap(log2(normalized_counts[top_genes, ] + 1),
        name="log2(normalized counts + 1)",
        top_annotation = ha,
        left_annotation = ra,
        show_row_names = TRUE,
        show_column_names = TRUE)

# dev.off()
```

------------------------------------------------------------------------

# HERE ABOVE ARE FOR PSET6. NO NEED TO LOOD DOWN

```{r}
geneset_oi = readr::read_tsv(url("https://zenodo.org/record/3260758/files/pemt_signature.txt"), col_names = "gene")%>% pull(gene) %>% .[. %in% rownames(ligand_target_matrix)] # only consider genes also present in the NicheNet model - this excludes genes from the gene list for which the official HGNC symbol was not used by Puram et al.

head(geneset_oi)
## [1] "SERPINE1" "TGFBI"    "MMP10"    "LAMC2"    "P4HA2"    "PDPN"

background_expressed_genes = expressed_genes_receiver %>% .[. %in% rownames(ligand_target_matrix)]
head(background_expressed_genes)
## [1] "RPS11"   "ELMO2"   "PNMA1"   "MMP2"    "TMEM216" "ERCC5"
```

# Step 3: Define a set of potential ligands

As potentially active ligands, we will use ligands that are 1) expressed by CAFs and 2) can bind a (putative) receptor expressed by malignant cells. Putative ligand-receptor links were gathered from NicheNet’s ligand-receptor data sources.

```{r}
# If wanted, users can remove ligand-receptor interactions that were predicted based on protein-protein interactions and only keep ligand-receptor interactions that are described in curated databases. To do this: uncomment following line of code:
# lr_network = lr_network %>% filter(database != "ppi_prediction_go" & database != "ppi_prediction")

ligands = lr_network %>% pull(from) %>% unique()
expressed_ligands = intersect(ligands,expressed_genes_sender)

receptors = lr_network %>% pull(to) %>% unique()
expressed_receptors = intersect(receptors,expressed_genes_receiver)

lr_network_expressed = lr_network %>% filter(from %in% expressed_ligands & to %in% expressed_receptors) 
head(lr_network_expressed)
## # A tibble: 6 × 2
##   from   to     
##   <chr>  <chr>  
## 1 A2M    MMP2   
## 2 A2M    MMP9   
## 3 ADAM10 APP    
## 4 ADAM10 CD44   
## 5 ADAM10 TSPAN5 
## 6 ADAM10 TSPAN15
```

This ligand-receptor network contains the expressed ligand-receptor interactions. As potentially active ligands for the NicheNet analysis, we will consider the ligands from this network.

```{r}
potential_ligands = lr_network_expressed %>% pull(from) %>% unique()
head(potential_ligands)
## [1] "A2M"    "ADAM10" "ADAM12" "ADAM15" "ADAM17" "ADAM9"
```

# Step 4: Perform NicheNet’s ligand activity analysis on the gene set of interest

Now perform the ligand activity analysis: in this analysis, we will calculate the ligand activity of each ligand, or in other words, we will assess how well each CAF-ligand can predict the p-EMT gene set compared to the background of expressed genes (predict whether a gene belongs to the p-EMT program or not).

```{r}
ligand_activities = predict_ligand_activities(geneset = geneset_oi, background_expressed_genes = background_expressed_genes, ligand_target_matrix = ligand_target_matrix, potential_ligands = potential_ligands)
```

Now, we want to rank the ligands based on their ligand activity. In our validation study, we showed that the area under the precision-recall curve (AUPR) between a ligand’s target predictions and the observed transcriptional response was the most informative measure to define ligand activity (this was the Pearson correlation for v1). Therefore, we will rank the ligands based on their AUPR. This allows us to prioritize p-EMT-regulating ligands.

```{r}
ligand_activities %>% arrange(-aupr_corrected) 
## # A tibble: 212 × 5
##    test_ligand auroc   aupr aupr_corrected pearson
##    <chr>       <dbl>  <dbl>          <dbl>   <dbl>
##  1 TGFB2       0.772 0.120          0.105    0.195
##  2 BMP8A       0.774 0.0852         0.0699   0.175
##  3 INHBA       0.777 0.0837         0.0685   0.122
##  4 CXCL12      0.714 0.0829         0.0676   0.141
##  5 LTBP1       0.727 0.0762         0.0609   0.160
##  6 CCN2        0.736 0.0734         0.0581   0.141
##  7 TNXB        0.719 0.0717         0.0564   0.157
##  8 ENG         0.764 0.0703         0.0551   0.145
##  9 BMP5        0.750 0.0691         0.0538   0.148
## 10 VCAN        0.720 0.0687         0.0534   0.140
## # … with 202 more rows
best_upstream_ligands = ligand_activities %>% top_n(30, aupr_corrected) %>% arrange(-aupr_corrected) %>% pull(test_ligand)
head(best_upstream_ligands)
## [1] "TGFB2"  "BMP8A"  "INHBA"  "CXCL12" "LTBP1"  "CCN2"
```

We see here that the performance metrics indicate that the 30 top-ranked ligands can predict the p-EMT genes reasonably, this implies that ranking of the ligands might be accurate as shown in our study. However, it is possible that for some gene sets, the target gene prediction performance of the top-ranked ligands would not be much better than random prediction. In that case, prioritization of ligands will be less trustworthy.

Additional note: we looked at the top 30 ligands here and will continue the analysis by inferring p-EMT target genes of these 30 ligands. However, the choice of looking only at the 30 top-ranked ligands for further biological interpretation is based on biological intuition and is quite arbitrary. Therefore, users can decide to continue the analysis with a different number of ligands. We recommend to check the selected cutoff by looking at the distribution of the ligand activity values. Here, we show the ligand activity histogram (the score for the 30th ligand is indicated via the dashed line).

```{r}
# show histogram of ligand activity scores
p_hist_lig_activity = ggplot(ligand_activities, aes(x=aupr_corrected)) + 
  geom_histogram(color="black", fill="darkorange")  + 
  # geom_density(alpha=.1, fill="orange") +
  geom_vline(aes(xintercept=min(ligand_activities %>% top_n(30, aupr_corrected) %>% pull(aupr_corrected))), color="red", linetype="dashed", linewidth=1) + 
  labs(x="ligand activity (PCC)", y = "# ligands") +
  theme_classic()
p_hist_lig_activity
```

# Step 5: Infer target genes of top-ranked ligands and visualize in a heatmap

Now we will show how you can look at the regulatory potential scores between ligands and target genes of interest. In this case, we will look at links between top-ranked p-EMT regulating ligands and p-EMT genes. In the ligand-target heatmaps, we show here regulatory potential scores for interactions between the 20 top-ranked ligands and following target genes: genes that belong to the gene set of interest and to the 250 most strongly predicted targets of at least one of the 20 top-ranked ligands (the top 250 targets according to the general prior model, so not the top 250 targets for this dataset). Consequently, genes of your gene set that are not a top target gene of one of the prioritized ligands, will not be shown on the heatmap.

```{r}
active_ligand_target_links_df = best_upstream_ligands %>% lapply(get_weighted_ligand_target_links,geneset = geneset_oi, ligand_target_matrix = ligand_target_matrix, n = 250) %>% bind_rows()

nrow(active_ligand_target_links_df)
## [1] 460
head(active_ligand_target_links_df)
## # A tibble: 6 × 3
##   ligand target  weight
##   <chr>  <chr>    <dbl>
## 1 TGFB2  ACTN1   0.0849
## 2 TGFB2  C1S     0.124 
## 3 TGFB2  COL17A1 0.0732
## 4 TGFB2  COL1A1  0.243 
## 5 TGFB2  COL4A2  0.148 
## 6 TGFB2  F3      0.0747
```

For visualization purposes, we adapted the ligand-target regulatory potential matrix as follows. Regulatory potential scores were set as 0 if their score was below a predefined threshold, which was here the 0.25 quantile of scores of interactions between the 30 top-ranked ligands and each of their respective top targets (see the ligand-target network defined in the data frame).

```{r}
active_ligand_target_links = prepare_ligand_target_visualization(ligand_target_df = active_ligand_target_links_df, ligand_target_matrix = ligand_target_matrix, cutoff = 0.25)

nrow(active_ligand_target_links_df)
## [1] 460
head(active_ligand_target_links_df)
## # A tibble: 6 × 3
##   ligand target  weight
##   <chr>  <chr>    <dbl>
## 1 TGFB2  ACTN1   0.0849
## 2 TGFB2  C1S     0.124 
## 3 TGFB2  COL17A1 0.0732
## 4 TGFB2  COL1A1  0.243 
## 5 TGFB2  COL4A2  0.148 
## 6 TGFB2  F3      0.0747
```

The putatively active ligand-target links will now be visualized in a heatmap. The order of the ligands accord to the ranking according to the ligand activity prediction.

```{r}
order_ligands = intersect(best_upstream_ligands, colnames(active_ligand_target_links)) %>% rev()
order_targets = active_ligand_target_links_df$target %>% unique()
vis_ligand_target = active_ligand_target_links[order_targets,order_ligands] %>% t()

p_ligand_target_network = vis_ligand_target %>% make_heatmap_ggplot("Prioritized CAF-ligands","p-EMT genes in malignant cells", color = "purple",legend_position = "top", x_axis_position = "top",legend_title = "Regulatory potential") + scale_fill_gradient2(low = "whitesmoke",  high = "purple", breaks = c(0,0.005,0.01)) + theme(axis.text.x = element_text(face = "italic"))

p_ligand_target_network
```

Note that the choice of these cutoffs for visualization is quite arbitrary. We recommend users to test several cutoff values.

If you would consider more than the top 250 targets based on prior information, you will infer more, but less confident, ligand-target links; by considering less than 250 targets, you will be more stringent.

If you would change the quantile cutoff that is used to set scores to 0 (for visualization purposes), lowering this cutoff will result in a more dense heatmap, whereas highering this cutoff will result in a more sparse heatmap.

# Follow-up analysis 1: Ligand-receptor network inference for top-ranked ligands

One type of follow-up analysis is looking at which receptors of the receiver cell population (here: malignant cells) can potentially bind to the prioritized ligands from the sender cell population (here: CAFs).

So, we will now infer the predicted ligand-receptor interactions of the top-ranked ligands and visualize these in a heatmap.

```{r}
# get the ligand-receptor network of the top-ranked ligands
lr_network_top = lr_network %>% filter(from %in% best_upstream_ligands & to %in% expressed_receptors) %>% distinct(from,to)
best_upstream_receptors = lr_network_top %>% pull(to) %>% unique()

# get the weights of the ligand-receptor interactions as used in the NicheNet model
weighted_networks = readRDS(url("https://zenodo.org/record/7074291/files/weighted_networks_nsga2r_final.rds"))
lr_network_top_df = weighted_networks$lr_sig %>% filter(from %in% best_upstream_ligands & to %in% best_upstream_receptors)

# convert to a matrix
lr_network_top_df = lr_network_top_df %>% spread("from","weight",fill = 0)
lr_network_top_matrix = lr_network_top_df %>% select(-to) %>% as.matrix() %>% magrittr::set_rownames(lr_network_top_df$to)

# perform hierarchical clustering to order the ligands and receptors
dist_receptors = dist(lr_network_top_matrix, method = "binary")
hclust_receptors = hclust(dist_receptors, method = "ward.D2")
order_receptors = hclust_receptors$labels[hclust_receptors$order]

dist_ligands = dist(lr_network_top_matrix %>% t(), method = "binary")
hclust_ligands = hclust(dist_ligands, method = "ward.D2")
order_ligands_receptor = hclust_ligands$labels[hclust_ligands$order]
```

Show a heatmap of the ligand-receptor interactions

```{r}
vis_ligand_receptor_network = lr_network_top_matrix[order_receptors, order_ligands_receptor]
p_ligand_receptor_network = vis_ligand_receptor_network %>% t() %>% make_heatmap_ggplot("Prioritized CAF-ligands","Receptors expressed by malignant cells", color = "mediumvioletred", x_axis_position = "top",legend_title = "Prior interaction potential")
p_ligand_receptor_network
```

# Follow-up analysis 2: Visualize expression of top-predicted ligands and their target genes in a combined heatmap

NicheNet only considers expressed ligands of sender cells, but does not take into account their expression for ranking the ligands. The ranking is purely based on the potential that a ligand might regulate the gene set of interest, given prior knowledge. Because it is also useful to further look into expression of ligands and their target genes, we demonstrate here how you could make a combined figure showing ligand activity, ligand expression, target gene expression and ligand-target regulatory potential.

Load additional packages required for the visualization:

```{r}
library(RColorBrewer)
library(cowplot)
library(ggpubr)
```

Prepare the ligand activity matrix

```{r}
ligand_aupr_matrix = ligand_activities %>% select(aupr_corrected) %>% as.matrix() %>% magrittr::set_rownames(ligand_activities$test_ligand)

vis_ligand_aupr = ligand_aupr_matrix[order_ligands, ] %>% as.matrix(ncol = 1) %>% magrittr::set_colnames("AUPR")
```

```{r}
p_ligand_aupr = vis_ligand_aupr %>% make_heatmap_ggplot("Prioritized CAF-ligands","Ligand activity", color = "darkorange",legend_position = "top", x_axis_position = "top", legend_title = "AUPR\n(target gene prediction ability)")
p_ligand_aupr
```

Prepare expression of ligands in fibroblast per tumor

Because the single-cell data was collected from multiple tumors, we will show here the average expression of the ligands per tumor.

```{r}
expression_df_CAF = expression[CAF_ids,order_ligands] %>% data.frame() %>% rownames_to_column("cell") %>% as_tibble() %>% inner_join(sample_info %>% select(cell,tumor), by =  "cell")

aggregated_expression_CAF = expression_df_CAF %>% group_by(tumor) %>% select(-cell) %>% summarise_all(mean)

aggregated_expression_df_CAF = aggregated_expression_CAF %>% select(-tumor) %>% t() %>% magrittr::set_colnames(aggregated_expression_CAF$tumor) %>% data.frame() %>% rownames_to_column("ligand") %>% as_tibble() 

aggregated_expression_matrix_CAF = aggregated_expression_df_CAF %>% select(-ligand) %>% as.matrix() %>% magrittr::set_rownames(aggregated_expression_df_CAF$ligand)

order_tumors = c("HN6","HN20","HN26","HN28","HN22","HN25","HN5","HN18","HN17","HN16") # this order was determined based on the paper from Puram et al. Tumors are ordered according to p-EMT score.
vis_ligand_tumor_expression = aggregated_expression_matrix_CAF[order_ligands,order_tumors]
```

```{r}
library(RColorBrewer)
color = colorRampPalette(rev(brewer.pal(n = 7, name ="RdYlBu")))(100)
p_ligand_tumor_expression = vis_ligand_tumor_expression %>% make_heatmap_ggplot("Prioritized CAF-ligands","Tumor", color = color[100],legend_position = "top", x_axis_position = "top", legend_title = "Expression\n(averaged over\nsingle cells)") + theme(axis.text.y = element_text(face = "italic"))
p_ligand_tumor_expression
```

Prepare expression of target genes in malignant cells per tumor

```{r}
expression_df_target = expression[malignant_ids,geneset_oi] %>% data.frame() %>% rownames_to_column("cell") %>% as_tibble() %>% inner_join(sample_info %>% select(cell,tumor), by =  "cell") 

aggregated_expression_target = expression_df_target %>% group_by(tumor) %>% select(-cell) %>% summarise_all(mean)

aggregated_expression_df_target = aggregated_expression_target %>% select(-tumor) %>% t() %>% magrittr::set_colnames(aggregated_expression_target$tumor) %>% data.frame() %>% rownames_to_column("target") %>% as_tibble() 

aggregated_expression_matrix_target = aggregated_expression_df_target %>% select(-target) %>% as.matrix() %>% magrittr::set_rownames(aggregated_expression_df_target$target)

vis_target_tumor_expression_scaled = aggregated_expression_matrix_target %>% t() %>% scale_quantile() %>% .[order_tumors,order_targets]
```

```{r}
p_target_tumor_scaled_expression = vis_target_tumor_expression_scaled  %>% make_threecolor_heatmap_ggplot("Tumor","Target", low_color = color[1],mid_color = color[50], mid = 0.5, high_color = color[100], legend_position = "top", x_axis_position = "top" , legend_title = "Scaled expression\n(averaged over\nsingle cells)") + theme(axis.text.x = element_text(face = "italic"))
p_target_tumor_scaled_expression
```

Combine the different heatmaps in one overview figure

```{r}
figures_without_legend = plot_grid(
  p_ligand_aupr + theme(legend.position = "none", axis.ticks = element_blank()) + theme(axis.title.x = element_text()),
  p_ligand_tumor_expression + theme(legend.position = "none", axis.ticks = element_blank()) + theme(axis.title.x = element_text()) + ylab(""),
  p_ligand_target_network + theme(legend.position = "none", axis.ticks = element_blank()) + ylab(""), 
  NULL,
  NULL,
  p_target_tumor_scaled_expression + theme(legend.position = "none", axis.ticks = element_blank()) + xlab(""), 
  align = "hv",
  nrow = 2,
  rel_widths = c(ncol(vis_ligand_aupr)+ 4.5, ncol(vis_ligand_tumor_expression), ncol(vis_ligand_target)) -2,
  rel_heights = c(nrow(vis_ligand_aupr), nrow(vis_target_tumor_expression_scaled) + 3)) 

legends = plot_grid(
  as_ggplot(get_legend(p_ligand_aupr)),
  as_ggplot(get_legend(p_ligand_tumor_expression)),
  as_ggplot(get_legend(p_ligand_target_network)),
  as_ggplot(get_legend(p_target_tumor_scaled_expression)),
  nrow = 2,
  align = "h")

plot_grid(figures_without_legend, 
          legends, 
          rel_heights = c(10,2), nrow = 2, align = "hv")
```

# Other follow-up analyses:

As another follow-up analysis, you can **infer possible signaling paths between ligands and targets of interest**. You can read how to do this in the following vignette [Inferring ligand-to-target signaling paths](https://github.com/saeyslab/nichenetr/blob/master/vignettes/ligand_target_signaling_path.md):vignette("ligand_target_signaling_path", package="nichenetr"). **(Jia: Could be very useful for Estrogen signaling and CCL8 signaling!!!)**

Another follow-up analysis is getting a “tangible” measure of how well top-ranked ligands predict the gene set of interest and assess which genes of the gene set can be predicted well. You can read how to do this in the following vignette [Assess how well top-ranked ligands can predict a gene set of interest](https://github.com/saeyslab/nichenetr/blob/master/vignettes/target_prediction_evaluation_geneset.md):vignette("target_prediction_evaluation_geneset", package="nichenetr").

In case you want to visualize ligand-target links between multiple interacting cells, you can make an appealing circos plot as shown in vignette [Circos plot visualization to show active ligand-target links between interacting cells](https://github.com/saeyslab/nichenetr/blob/master/vignettes/circos.md):vignette("circos", package="nichenetr").

# References

**NicheNet**

Browaeys, R.; Saelens, W.; Saeys, Y. NicheNet: Modeling Intercellular Communication by Linking Ligands to Target Genes. Nat. Methods 2020, 17 (2), 159–162. <https://doi.org/10.1038/s41592-019-0667-5>.

**RNA-seq data of fat tissue resident macrophages in fat (VAMs) and other resident cells**

Moura Silva Hernandez; Kitoko Jamil Zola; Queiroz Camila Pereira; Kroehling Lina; Matheis Fanny; Yang Katharine Lu; Reis Bernardo S.; Ren-Fielding Christine; Littman Dan R.; Bozza Marcelo Torres; Mucida Daniel; Lafaille Juan J. C-MAF–Dependent Perivascular Macrophages Regulate Diet-Induced Metabolic Syndrome. Sci. Immunol. 2021, 6 (64), eabg7506. [https://doi.org/10.1126/sciimmunol.abg7506.![image](https://github.com/Jia-Zhao1998/JiaPSET6/assets/67213486/3a65aea6-b3d7-4e45-9b02-ee70328722ec)](https://doi.org/10.1126/sciimmunol.abg7506.!%5Bimage%5D(https://github.com/Jia-Zhao1998/JiaPSET6/assets/67213486/3a65aea6-b3d7-4e45-9b02-ee70328722ec))
